File: app/src/main/java/com/fanchen/imovie/base/BaseRecyclerActivity.java
package com.fanchen.imovie.base;

import android.os.Bundle;
import android.support.v4.widget.SwipeRefreshLayout;
import android.support.v7.widget.RecyclerView;
import android.util.TypedValue;
import android.view.LayoutInflater;
import android.view.View;

import com.fanchen.imovie.R;
import com.fanchen.imovie.retrofit.callback.RefreshCallback;
import com.fanchen.imovie.retrofit.RetrofitManager;
import com.fanchen.imovie.thread.task.AsyTaskListener;
import com.fanchen.imovie.view.CustomEmptyView;
import com.squareup.picasso.Picasso;

import butterknife.InjectView;

/**
 * layout为R.layout.activity_recyclerview
 * 以RecyclerView为显示根布局的activity继承该类
 * Created by fanchen on 2017/8/22.
 */
public abstract class BaseRecyclerActivity extends BaseToolbarActivity implements
        View.OnClickListener, SwipeRefreshLayout.OnRefreshListener, BaseAdapter.OnItemClickListener, BaseAdapter.OnLoadListener {

    @InjectView(R.id.recycle_list)
    protected RecyclerView mRecyclerView;
    @InjectView(R.id.swipe_refresh_layout)
    protected SwipeRefreshLayout mSwipeRefreshLayout;
    @InjectView(R.id.cev_empty)
    protected CustomEmptyView mCustomEmptyView;

    private int page = 1;
    private int pageStart = 1;
    private BaseAdapter mAdapter;

    @Override
    protected int getLayout() {
        return R.layout.activity_recyclerview;
    }

    @Override
    protected void initActivity(Bundle savedState, LayoutInflater inflater) {
        super.initActivity(savedState, inflater);
        TypedValue typedValue = new TypedValue();
        getTheme().resolveAttribute(R.attr.colorPrimary, typedValue, true);
        if (!checkFieldNull()) {
            mSwipeRefreshLayout.setColorSchemeColors(typedValue.data);
            mRecyclerView.setLayoutManager(getLayoutManager());
            mRecyclerView.setAdapter(mAdapter = getAdapter(getPicasso()));
        }
        loadData(getRetrofitManager(), page = pageStart);
    }

    @Override
    protected void setListener() {
        super.setListener();
        if (!checkFieldNull()) {
            mSwipeRefreshLayout.setOnRefreshListener(this);
            mCustomEmptyView.setOnClickListener(this);
            mRecyclerView.addOnScrollListener(scrollListener);
        }
        if (mAdapter != null) {
            mAdapter.setOnItemClickListener(this);
            if (hasLoad()) mAdapter.setOnLoadListener(this);
        }
    }

    public boolean checkFieldNull() {
        return mSwipeRefreshLayout == null || mCustomEmptyView == null || mRecyclerView == null;
    }

    @Override
    public void onClick(View v) {
        switch (v.getId()) {
            case R.id.iv_top_back:
                finish();
                break;
            case R.id.cev_empty:
                onRefresh();
                break;
        }
    }

    @Override
    public void onRefresh() {
        loadData(getRetrofitManager(), page = pageStart);
    }

    @Override
    public void onLoad() {
        loadData(getRetrofitManager(), ++page);
    }

    public int getPage() {
        return page;
    }

    public void setPageStart(int page) {
        this.pageStart = page;
    }

    public void setPage(int page) {
        this.page = page;
    }

    /**
     * @return
     */
    protected boolean isRefresh() {
        return page <= pageStart;
    }

    /**
     * @return
     */
    protected boolean hasLoad() {
        return false;
    }

    /**
     * @return
     */
    protected abstract RecyclerView.LayoutManager getLayoutManager();

    /**
     * @return
     */
    protected abstract BaseAdapter getAdapter(Picasso picasso);

    /**
     * @param page
     */
    protected abstract void loadData(RetrofitManager retrofit, int page);

    /**
     *
     */
    protected RecyclerView.OnScrollListener scrollListener = new RecyclerView.OnScrollListener() {

        @Override
        public void onScrollStateChanged(RecyclerView recyclerView, int newState) {
            super.onScrollStateChanged(recyclerView, newState);
            Picasso picasso = getPicasso();
            if (picasso != null) {
                if (newState == RecyclerView.SCROLL_STATE_IDLE) {
                    picasso.resumeTag(BaseRecyclerActivity.class);
                } else {
                    picasso.pauseTag(BaseRecyclerActivity.class);
                }
            }
        }

    };

    /**
     * @param <T>
     */
    protected abstract class RefreshRecyclerActivityImpl<T> implements RefreshCallback<T> {

        @Override
        public void onStart(int enqueueKey) {
            if (mCustomEmptyView == null || mSwipeRefreshLayout == null || mAdapter == null) return;
            mCustomEmptyView.setEmptyType(CustomEmptyView.TYPE_NON);
            if (!mSwipeRefreshLayout.isRefreshing() && !mAdapter.isLoading()) {
                mSwipeRefreshLayout.setRefreshing(true);
            }
        }

        @Override
        public void onFailure(int enqueueKey, String throwable) {
            if (mCustomEmptyView == null || mAdapter == null) return;
            if (mAdapter.isEmpty()) mCustomEmptyView.setEmptyType(CustomEmptyView.TYPE_ERROR);
            showSnackbar(throwable);
        }

        @Override
        public void onFinish(int enqueueKey) {
            if (mCustomEmptyView == null || mSwipeRefreshLayout == null || mAdapter == null) return;
            mSwipeRefreshLayout.setRefreshing(false);
            mAdapter.setLoading(false);
            if (mAdapter.isEmpty()) {
                mCustomEmptyView.setEmptyType(CustomEmptyView.TYPE_EMPTY);
            }
        }
    }

    protected abstract class TaskRecyclerActivityImpl<T> implements AsyTaskListener<T> {

        @Override
        public void onTaskFinish() {
            if (mCustomEmptyView == null || mSwipeRefreshLayout == null || mAdapter == null) return;
            mSwipeRefreshLayout.setRefreshing(false);
            mAdapter.setLoading(false);
            if (mAdapter.isEmpty()) {
                mCustomEmptyView.setEmptyType(CustomEmptyView.TYPE_EMPTY);
            }
        }

        @Override
        public void onTaskSart() {
            if (mCustomEmptyView == null || mSwipeRefreshLayout == null) return;
            mCustomEmptyView.setEmptyType(CustomEmptyView.TYPE_NON);
            if (!mSwipeRefreshLayout.isRefreshing() && !mAdapter.isLoading()) {
                mSwipeRefreshLayout.setRefreshing(true);
            }
        }

        @Override
        public void onTaskProgress(Integer... values) {

        }

    }
}

File: app/src/main/java/com/fanchen/imovie/base/BaseRecyclerActivity.kt
package com.fanchen.imovie.base

import android.os.Bundle
import android.support.v4.widget.SwipeRefreshLayout
import android.support.v7.widget.RecyclerView
import android.util.TypedValue
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import com.fanchen.imovie.R
import com.fanchen.imovie.retrofit.RetrofitManager
import com.fanchen.imovie.retrofit.callback.RefreshCallback
import com.fanchen.imovie.thread.task.AsyTaskListener
import com.fanchen.imovie.view.CustomEmptyView
import com.squareup.picasso.Picasso
import kotlinx.android.synthetic.main.activity_recyclerview.*

/**
 * layout为R.layout.activity_recyclerview
 * 以RecyclerView为显示根布局的activity继承该类
 */
abstract class BaseRecyclerActivity : BaseToolbarActivity(), View.OnClickListener, SwipeRefreshLayout.OnRefreshListener, BaseAdapter.OnItemClickListener, BaseAdapter.OnLoadListener {

    var page = 1
    private var pageStart = 1
    private var mAdapter: BaseAdapter? = null

    /**
     * @return
     */
    protected val isRefresh: Boolean
        get() = page <= pageStart

    /**
     * @return
     */
    protected abstract val layoutManager: RecyclerView.LayoutManager

    lateinit var mTextView: TextView
    var mRecyclerView: RecyclerView? = null
    var mSwipeRefreshLayout: SwipeRefreshLayout? = null
    lateinit var mCustomEmptyView: CustomEmptyView

    /**
     *
     */
    protected var scrollListener: RecyclerView.OnScrollListener = object : RecyclerView.OnScrollListener() {

        override fun onScrollStateChanged(recyclerView: RecyclerView?, newState: Int) {
            super.onScrollStateChanged(recyclerView, newState)
            val picasso = picasso
            if (picasso != null) {
                if (newState == RecyclerView.SCROLL_STATE_IDLE) {
                    picasso.resumeTag(BaseRecyclerActivity::class.java)
                } else {
                    picasso.pauseTag(BaseRecyclerActivity::class.java)
                }
            }
        }

    }

    override fun getLayout(): Int {
        return R.layout.activity_recyclerview
    }

    override fun initActivity(savedState: Bundle?, inflater: LayoutInflater) {
        super.initActivity(savedState, inflater)
        val typedValue = TypedValue()
        theme.resolveAttribute(R.attr.colorPrimary, typedValue, true)
        if (!checkFieldNull()) {
            swipe_refresh_layout!!.setColorSchemeColors(typedValue.data)
            recycle_list.layoutManager = layoutManager
            recycle_list.adapter = getAdapter(picasso)
        }
        loadData(retrofitManager, page = pageStart)
    }

    override fun setListener() {
        super.setListener()
        if (!checkFieldNull()) {
            swipe_refresh_layout!!.setOnRefreshListener(this)
            cev_empty!!.setOnClickListener(this)
            recycle_list.addOnScrollListener(scrollListener)
        }
        if (mAdapter != null) {
            mAdapter!!.setOnItemClickListener(this)
            if (hasLoad()) mAdapter!!.setOnLoadListener(this)
        }
    }

    private fun checkFieldNull(): Boolean {
        return swipe_refresh_layout == null || cev_empty == null || recycle_list == null
    }

    override fun onClick(v: View) {
        when (v.id) {
            R.id.iv_top_back -> finish()
            R.id.cev_empty -> onRefresh()
        }
    }

    override fun onRefresh() {
        loadData(retrofitManager, page = pageStart)
    }

    override fun onLoad() {
        loadData(retrofitManager, ++page)
    }

    fun setPageStart(page: Int) {
        this.pageStart = page
    }

    /**
     * @return
     */
    protected open fun hasLoad(): Boolean {
        return false
    }

    /**
     * @return
     */
    protected abstract fun getAdapter(picasso: Picasso?): BaseAdapter

    /**
     * @param page
     */
    protected abstract fun loadData(retrofit: RetrofitManager, page: Int)

    /**
     * @param <T>
    </T> */
    protected abstract inner class RefreshRecyclerActivityImpl<T> : RefreshCallback<T> {

        override fun onStart(enqueueKey: Int) {
            if (cev_empty == null || swipe_refresh_layout == null || mAdapter == null) return
            cev_empty!!.setEmptyType(CustomEmptyView.TYPE_NON)
            if (!swipe_refresh_layout!!.isRefreshing && !mAdapter!!.isLoading) {
                swipe_refresh_layout!!.isRefreshing = true
            }
        }

        override fun onFailure(enqueueKey: Int, throwable: String) {
            if (cev_empty == null || mAdapter == null) return
            if (mAdapter!!.isEmpty) cev_empty!!.setEmptyType(CustomEmptyView.TYPE_ERROR)
            showSnackbar(throwable)
        }

        override fun onFinish(enqueueKey: Int) {
            if (cev_empty == null || swipe_refresh_layout == null || mAdapter == null) return
            swipe_refresh_layout!!.isRefreshing = false
            mAdapter!!.isLoading = false
            if (mAdapter!!.isEmpty) {
                cev_empty!!.setEmptyType(CustomEmptyView.TYPE_EMPTY)
            }
        }
    }

    protected abstract inner class TaskRecyclerActivityImpl<T> : AsyTaskListener<T> {

        override fun onTaskFinish() {
            if (cev_empty == null || swipe_refresh_layout == null || mAdapter == null) return
            swipe_refresh_layout!!.isRefreshing = false
            mAdapter!!.isLoading = false
            if (mAdapter!!.isEmpty) {
                cev_empty!!.setEmptyType(CustomEmptyView.TYPE_EMPTY)
            }
        }

        override fun onTaskStart() {
            if (cev_empty == null || swipe_refresh_layout == null) return
            cev_empty!!.setEmptyType(CustomEmptyView.TYPE_NON)
            if (!swipe_refresh_layout!!.isRefreshing && !mAdapter!!.isLoading) {
                swipe_refresh_layout!!.isRefreshing = true
            }
        }
    }
}
